### PR TITLE
chore: TypeScript project references

### PR DESCRIPTION
**Pull request by Tohaker**
Project References in our app make it faster to develop multiple packages at once. For example, consider you want to create a function in the `time` package and consume it in `ui`; here are the current and proposed workflows;

**Current**
1. Add your function and tests to `time`
2. Run `pnpm build` over the whole app (~30s on modern machines)
3. Consume the function in `ui` (which may also require restarting your IDE's TS Server)

**Proposed**
1. Add your function and tests to `time`
2. Consume the function in `ui` (the IDE TS Server won't need to be restarted)

This should make the developer experience a lot faster and simpler

### CODE REVIEW

**Comment by Tohaker** [comment]
@danstarns @kreshnnikgashi An odd thing happened on this PR... [this test](https://github.com/rocket-connect/graphql-debugger/actions/runs/6213112127/job/16863867097#step:5:185) failed but the workflow still "passed". Any idea why the `pnpm` command didn't return a non-zero exit code?

**Comment by danstarns** [comment]
@Tohaker thanks for raising this, I think it has something to do with this addition the other day: 

* https://github.com/rocket-connect/graphql-debugger/commit/cac071b244816e5b631b299cd8809509277986f2#diff-b31f1e5654d027d84c24cdba2ff31ab79e3ea617091ed348a97476a27b076078R20

I added some more tests to the mono repo, and there were race conditions using turbo repo as it runs all the tests in parallel.

It has to do with this line: 

``` 
pnpm recursive exec -- sh -c 'pnpm run test || true'
``` 

```
|| true
```

I have made a PR error to throw if they fail: 

* https://github.com/rocket-connect/graphql-debugger/pull/49

File: .github/actions/setup-project/action.yml [hunk 1]
```diff
@@ -7,9 +7,6 @@ inputs:
   node-version:
     required: false
     default: 18
-  turbo-version:
```
Comment (danstarns) [inline .github/actions/setup-project/action.yml:10]: Does removing this void any cache it may store?

File: e2e/jest.config.js [hunk 2]
```diff
@@ -1,3 +1,7 @@
+const { compilerOptions } = require("./tsconfig.json");
+const { pathsToModuleNameMapper } = require("ts-jest");
+
+/** @type {import('ts-jest').JestConfigWithTsJest} */
```
Comment (danstarns) [inline e2e/jest.config.js:4]: :)

File: e2e/package.json [hunk 3]
```diff
@@ -12,14 +12,13 @@
   },
   "homepage": "https://github.com/rocket-connect/graphql-debugger#readme",
   "scripts": {
-    "build": "tsc",
+    "build": "tsc -b",
```
Comment (danstarns) [inline e2e/package.json:15]: Just to clarify what the -b is for here?

File: packages/collector-proxy/package.json [hunk 4]
```diff
@@ -13,26 +13,20 @@
...
   },
   "peerDependencies": {
     "graphql": "^16.0.0 || ^17.0.0"
   },
   "dependencies": {
-    "@graphql-debugger/time": "workspace:^",
-    "@graphql-debugger/utils": "workspace:^",
     "@graphql-debugger/data-access": "workspace:^",
     "@graphql-debugger/trace-schema": "workspace:^",
-    "@graphql-debugger/queue": "workspace:^",
```
Comment (danstarns) [inline packages/collector-proxy/package.json:27]: What happens when we push to npm, will it say these packages are not found?

**Review by danstarns — COMMENTED** [review]
This will really help with dev x, I think we need to think about releasing with this. 

I see lots of dependencies removed from the package JSON files and I wonder what happens when we push to npm and people try to download.

File: packages/graphql-schema/tsconfig.json [hunk 5]
```diff
@@ -2,15 +2,26 @@
   "extends": "@tsconfig/node18/tsconfig.json",
   "compilerOptions": {
     "target": "ES2020",
-    "esModuleInterop": true,
-    "forceConsistentCasingInFileNames": true,
-    "strict": true,
```
Comment (danstarns) [inline packages/graphql-schema/tsconfig.json:7]: 👀

I dont mind any here and there but do we really wanna be removing this if the project is already compiling?

File: packages/schemas/tsconfig.json [hunk 6]
```diff
@@ -2,12 +2,15 @@
   "extends": "@tsconfig/node18/tsconfig.json",
   "compilerOptions": {
     "target": "ES2020",
-    "esModuleInterop": true,
-    "forceConsistentCasingInFileNames": true,
-    "strict": true,
-    "skipLibCheck": true,
     "outDir": "build",
+    "rootDir": "./src",
```
Comment (danstarns) [inline packages/schemas/tsconfig.json:6]: Making sure this doing effect the "main" entry in the package json file

File: packages/time/src/unix-nano.ts [hunk 7]
```diff
@@ -19,6 +19,10 @@ export class UnixNanoTimeStamp {
     this.input = input;
   }
 
+  public static newFn() {
+    return 1;
```
Comment (danstarns) [inline packages/time/src/unix-nano.ts:23]: ?

File: packages/ui/package.json [hunk 8]
```diff
@@ -24,14 +24,11 @@
   },
   "dependencies": {
     "@fontsource/roboto": "5.0.8",
-    "@graphql-debugger/time": "workspace:^",
```
Comment (danstarns) [inline packages/ui/package.json:27]: ditto

**Review by danstarns — COMMENTED** [review]
Some more comments:

File: e2e/package.json (hunk 3 above)
Comment (Tohaker) [reply e2e/package.json:15]: That's the `build` option - you need it to ensure that all dependencies below this package are built. I found that turbo wasn't running anything in a particular order so it required this

File: .github/actions/setup-project/action.yml (hunk 1 above)
Comment (Tohaker) [reply .github/actions/setup-project/action.yml:10]: I don't think this was necessary, as `pnpm` installs it for us? And I believe the packages are cached already in our node setup action. That said, I'm no Turbo expert

File: packages/collector-proxy/package.json (hunk 4 above)
Comment (Tohaker) [reply packages/collector-proxy/package.json:27]: Do these packages get shipped to `npm` too? I thought they'd be internal and just packaged into the final build

File: packages/graphql-schema/tsconfig.json (hunk 5 above)
Comment (Tohaker) [reply packages/graphql-schema/tsconfig.json:7]: They're already included verbatim in the `node18` config! I'm not removing any configuration here

File: packages/time/src/unix-nano.ts (hunk 7 above)
Comment (Tohaker) [reply packages/time/src/unix-nano.ts:23]: Oops that's me testing stuff, will remove

**Comment by danstarns** [comment]
@Tohaker closing this now due to inactivity. Thank you for the idea an effort, reach out if you have more ideas or want to continue on this .

**Comment by Tohaker** [comment]
Sorry Dan, been super busy on my end. Hope all is going well with the rest of the project

**Pull request closed without merging**